roaring/roaring.go

```diff
@@ -705,93 +674,98 @@ func (b *Bitmap) unionInPlace(others ...*Bitmap) {
-			// so we need to calculate a union.
+			// determine whether we have a target to union into.
+			tContainer := target.Containers.Get(iKey)
+			// if the target's full, short-circuit out.
+			if tContainer != nil {
+				if tContainer.n == maxContainerVal+1 {
+					bitmapIters.markItersWithCurrentKeyAsHandled(i, iKey)
+					continue
+				}
+				expectedN = int64(tContainer.n)
+			}
+			// Check i and later iters for any max-range containers, and
+			// find out how many there are.
+			summaryStats := bitmapIters[i:].calculateSummaryStats(iKey)
 			if summaryStats.hasMaxRange {
 				// One (or more) of the containers represented the maximum possible
 				// range that a container can store, so instead of calculating a
 				// union we can generate an RLE container that represents the entire
 				// range.
-				container := &Container{
+				tContainer = &Container{
 					runs:          []interval16{{start: 0, last: maxContainerVal}},
 					containerType: containerRun,
 					n:             maxContainerVal + 1,
 				}
-				target.Containers.Put(iKey, container)
+				target.Containers.Put(iKey, tContainer)
 				bitmapIters.markItersWithCurrentKeyAsHandled(i, iKey)
 				continue
 			}
-
-			// Use a bitmap container for the target bitmap, and union everything
-			// into it.
-			//
-			// TODO(rartoul): Add another conditional case for n < ArrayMaxSize
-			// (or some fraction of that value) to avoid allocating expensive
-			// bitmaps when unioning many low-density array containers, but this
-			// will require writing a union in place algorithm for an array container
-			// that accepts multiple different containers to union into it for
-			// efficiency.
-			container := target.Containers.Get(iKey)
-			// If target already has a bitmap container for iKey then we can reuse that,
-			// otherwise we have to allocate a new one or convert the existing container
-			// into a bitmap container.
-			if container == nil {
-				buf := make([]uint64, bitmapN)
-				ob := buf[:bitmapN]
-				container = &Container{
-					bitmap:        ob,
-					n:             0,
-					containerType: containerBitmap,
+			expectedN += summaryStats.n
+			var itersToUnion handledIters
+			if tContainer == nil {
+				if summaryStats.c == 1 {
+					// There's no target and we have only one container, we
+					// can just clone it instead of unioning
+					statsHit("unionInPlace/reuse")
+					target.Containers.Put(iKey, iContainer.Clone())
+					bitmapIters[i].handled = true
+					continue
+				}
+				// If we expect enough items that we'll want to convert to bitmap,
+				// and the container we already found isn't a bitmap, make a bitmap
+				// and union everything into that. Otherwise clone the first
+				// container and union everything else into it.
+				if expectedN >= 512 && iContainer.containerType != containerBitmap {
+					statsHit("unionInPlace/newBitmap")
+					tContainer = &Container{containerType: containerBitmap, bitmap: make([]uint64, bitmapN)}
```

Comment:
could you just do:

```golang
tContainer = iContainer.toBitmap()
```
and then both branches can share the

```golang
itersToUnion = bitmapIters[i+1:]
```

line? Or is that not equivalent

Reply:
We don't currently have a toBitmap(), I don't think? So we have a "convert this container in-place to a bitmap", and we have a Clone(). It might make more sense to add a CloneAsBitmap(), though.